raco/compile.py

```diff
@@ -35,31 +37,39 @@ def write_if_enabled(self, plan, title):
         self.ind += 1
 
 
-def optimize_by_rules(expr, rules):
+def optimize_by_rules(expr, rules_lists):
     writer = PlanWriter()
     writer.write_if_enabled(expr, "before rules")
 
-    for rule in rules:
-        def recursiverule(e):
-            newe = rule(e)
-            writer.write_if_enabled(newe, str(rule))
-            if newe.stop_recursion:
-                return newe
-
-            # log the optimizer step
-            if str(e) == str(newe):
-                LOG.debug("apply rule %s (no effect)\n" +
-                          " %s \n", rule, e)
-            else:
-                LOG.debug("apply rule %s\n" +
-                          colored("  -", "red") + " %s" + "\n" +
-                          colored("  +", "green") + " %s", rule, e, newe)
-
-            newe.apply(recursiverule)
-
-            return newe
-        expr = recursiverule(expr)
-
+    for rules in rules_lists:
+        for i in range(20):
```

Comment:
If this lands in an infinite recursion, it'll blow the python stack soon enough, right? Do we really need an explicit cutoff?